pgcli/pgexecute.py

```diff
@@ -186,15 +186,15 @@ class PGExecute(object):
     version_query = "SELECT version();"
 
     def __init__(self, database, user, password, host, port, dsn, **kwargs):
-        self.dbname = database
-        self.user = user
-        self.password = password
-        self.host = host
-        self.port = port
-        self.dsn = dsn
-        self.extra_args = {k: unicode2utf8(v) for k, v in kwargs.items()}
+        self.conn = None
+        self.dbname = None
+        self.user = None
+        self.password = None
+        self.host = None
+        self.port = None
+        self.dsn = None
         self.server_version = None
-        self.connect()
+        self.connect(database, user, password, host, port, dsn, **kwargs)
```

Comment:
Why this change?

Reply:
It looked to me quite strange that we:
1. set properties at `__init__`
2. than pass it to `connect` method that also accepts all the arguments or takes if from object properties
3. and finally in `connect` sometimes sets properties at the end of it execution

I think it would be much easier to get the data flow if we set all the parameters at `__init__` method and use them in the `connect` method leaving it without parameters or if we don't set anything in the `__init__` and pass all the parameters to the `connect`.

What do you think of this change? If I didn't convince you I'll but the current logic back.

Reply:
The current pgexecute does what you describe in 1st scenario. It sets class member variables in `__init__`, and then calls `connect()` without parameters:

https://github.com/dbcli/pgcli/blob/7b03f8e2043682aa2d1c06e0f7e8258b2f9c43c5/pgcli/pgexecute.py#L188-L197

Your version conforms to the 2nd scenario (we don't set anything in `__init__` and pass all the parameters to `connect`), so I was wondering why this is better?

The only difference I see between the two is that in 2nd scenario, if the initial connection fails, you can't call `connect()` without parameters to retry. Otherwise it's pretty much the same.

Reply:
I think I could miss the point of what `PGExecute` properties are used for? Which of cases is more relevant, use them as connection parameters to establish new one or represent the state of an already established connection?

> The current pgexecute does what you describe in 1st scenario. It sets class member variables in __init__, and then calls connect() without parameters:

Yes, but it does it twice: once at the `__init__` and second at the `connect`. When I first read the code it was harder for me to predict behaviour of those 2 methods. Which of them sets the state of an object and which initiates the connection? If i don't pass `database` parameter  to `connect` what value would be used for connection? Now it's quite clear to me, but that was the first impression :-)

> Your version conforms to the 2nd scenario (we don't set anything in __init__ and pass all the parameters to connect), so I was wondering why this is better?

The idea was also to prepare to get the actual values of connection for the lower psycopg API (as v2.8 is released). What should we put into the `host` property of an `PGExecute` in case of multiple hosts? The original multi-host string or the final (the host we have connected to? What if we connected via DSN string should we reconnect using the same DSN string or using the parameters we parsed from it?

> if the initial connection fails, you can't call connect() without parameters to retry

I think it depends on design of `PGExecute`. If it stores initial values - it can reconnect itself, otherwise we need to store initial parameters somewhere outside.